### PR TITLE
Optimize Activity Ids

### LINKED ISSUE #46818: Faster Diagnostics Activity ID generation from random bytes
Diagnostics.Activitiy.Start method currently calls Guid.NewGuid to generate a new activity ID. However, GUID generation seems to work pretty slow on Linux, thus it's proposed to replace it with some random number generation (e.g. Random.Next()) which seems to work much faster.

### PR DESCRIPTION
**Pull request by tarekgh**
Fixes https://github.com/dotnet/runtime/issues/46818

Activity class uses `Guid.NewGuid()` to generate the Trace and Span Id's. It is proven on Linux Guid generation is very slow https://github.com/dotnet/runtime/issues/13628. It was noticed in some activity scenarios that Guid generation inside the Activity consuming none-trivial amount of time.
The fix here is to change the way we generate the Ids to use Random class instead of Guid.

# Linux

*Before*

|                Method |     Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------------- |---------:|----------:|----------:|-------:|------:|------:|----------:|
| ActivityIdsGeneration | 4.386 us | 0.0330 us | 0.0258 us | 0.0992 |     - |     - |     632 B |

*After*

|                Method |     Mean |   Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------------- |---------:|--------:|--------:|-------:|------:|------:|----------:|
| ActivityIdsGeneration | 305.9 ns | 4.71 ns | 4.17 ns | 0.0625 |     - |     - |     392 B |

# Windows

*Before*

|                Method |     Mean |   Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------------- |---------:|--------:|--------:|-------:|------:|------:|----------:|
| ActivityIdsGeneration | 544.1 ns | 6.12 ns | 5.72 ns | 0.1001 |     - |     - |     632 B |

*After*

|                Method |     Mean |   Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------------- |---------:|--------:|--------:|-------:|------:|------:|----------:|
| ActivityIdsGeneration | 355.5 ns | 5.28 ns | 4.94 ns | 0.0625 |     - |     - |     392 B |

### CODE REVIEW

**Comment by tarekgh** [comment]
CC @alnikola @davidfowl @noahfalk @stephentoub @ManickaP @sebastienros

**Comment by tarekgh** [comment]
Hold your review as I am looking at some issue on NetFX. Looks the Ids not generated correctly.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs
Comment (am11) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:1667]: It could probably be made simpler by calling the interop method directly. Something like:

```c#
internal static unsafe void SetToRandomBytes(Span<byte> outBytes)
{
    Debug.Assert(outBytes.Length == 16 || outBytes.Length == 8);
    Interop.GetRandomBytes((byte*)outBytes.GetPinnableReference(), outBytes.Length);
}
```

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:1667]: @am11 Using the interop will require to include the source file `Interop.Random.cs` to the project. That will require to build the assembly differently for OSX, Linux and Windows which we don't currently do. That will increase the package size too as we'll include more cross generated assemblies.

Comment (am11) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:1667]: Ah ok. I didn't looked at the proj file of this assmebly.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs
Comment (davidfowl) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:42]: I don't think random is thread safe https://github.com/dotnet/runtime/issues/43887.

Comment (pjanotti) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:42]: Hi @tarekgh - sorry if my memory betrays me but this seems to me that will cause collisions of trace IDs when using the package on .NET Framework: the seed is the time and it has a resolution around 15ms and reaches zero in a bit less than 50 days. Anytime that we get the same seed we will be repeating the trace ids that right now have a very strong guarantee of uniqueness.

A thing that perhaps also changed was that it was problematic to use the same random object concurrently, has that changed?

In order to keep the same strong guarantee for trace IDs, I think we need to keep using `NewGuid` for them. That said the span IDs don't need the same strong guarantees and perhaps can leverage `Random`.

Comment (stephentoub) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:42]: It's not. (Often this manifests as zeros starting to be returned from all accesses after some period of time / some number of collisions.)

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:42]: Is there a safe way to generate random numbers then? or should I look at different solution here?

Comment (pjanotti) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:42]: Refreshing my own memory: I looked at guaranteeing uniqueness about 6 months ago in the context of SignalFx instrumentation, ptal at the top comment https://github.com/signalfx/signalfx-dotnet-tracing/pull/64#issue-475043942

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:42]: Hi @pjanotti, thanks for the feedback. Yes I agree the current solution is not enough. `NewGuid` is very slow too which we are trying to change use something else. I'll look more and see if I can find better way.

Comment (stephentoub) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:42]: One typical solution employed is to have a ThreadStatic Random instance.  Or if you don't expect a lot of contention, you can also lock around access to the instance.

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:42]: >  (Often this manifests as zeros starting to be returned from all accesses after some period of time / some number of collisions.)

Is this generic issue in Random class? or this happen only with calling it from multiple threads in same time?

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:42]: @pjanotti what you think of the idea using thread static for Random and create it with a different seed for different threads. (see can be obtained from Guid.NewGuid().GetHashCode() as example). I can try that and measure it to see if it still be faster.

Comment (noahfalk) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:42]: I would guess the cost of a thread-static access is small relative to the cost of the random number generation. Haven't benchmarked myself but I am guess single-digit ns for thread-static read, double-digit ns for creating random data.

> Is this generic issue in Random class? or this happen only with calling it from multiple threads in same time?

I certainly hope it is only an issue when being unsafely shared for multi-threaded use, otherwise it sounds like Random would be quite broken.

Comment (stephentoub) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:42]: > Is this generic issue in Random class? or this happen only with calling it from multiple threads in same time? 

When erroneously accessed by multiple threads concurrently.

Comment (noahfalk) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:42]: Haven't thought it through, but a few initial thoughts:
a) we could make an internal version of Random that uses long instead of int for the seed, all the internal state, and the output type of the Next() function. This gets us to 63 bits of entropy (the implementation does Math.Abs() that loses 1 bit) and also doubles the throughput of the generator. The downside is maybe 100 mostly copied lines of code and double the amount of memory used by the per-thread Random.
b) Random uses a substractive generator that has an internal state that is 55x larger than what we are using for the seed. I don't have quick access to Knuth's book where he analyzes the algorithm but it appeared that if we put more entropy into that initialization then we get more entropy in the resulting random sequences. 
c) If the default seeds are weak we could always use something like Guid.NewGuid() to generate a strong random seed. As long as we only call it once per thread 1us should be neglible.

**Comment by noahfalk** [comment]
Is the code for that benchmark you ran shown somewhere @tarekgh? Is it measuring the total time to create, start, and stop an Activity? If the benchmark is only measuring the time to generate 16 bytes of random data I would hope we can do that in 20-50ns if we are being efficient.

**Comment by tarekgh** [comment]
@noahfalk the code I used is 

```C#
        [Benchmark] public int ActivityIdsGeneration()
        {
            using Activity activity = new Activity("MyActivity");
            activity.Start();
            ActivityTraceId traceId = activity.TraceId;
            ActivitySpanId spanId = activity.SpanId;

            return 100;
        }
```

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs
Comment (noahfalk) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:42]: > @pjanotti what you think of the idea using thread static for Random

If the size of the seed is 32 bits I think it would still cause trouble. I am assuming we have a bunch of VMs in a distributed system (~1000 machines), each of which is starting potentially many threads, and each of these threads starts at least one Activity. All of these Activities are being logged to a database and the transactions are stored for a few months before being deleted. If each machine during that time period created 100 threads (seems very likely for an that app restarts periodically) then there are 100K total seeds generated. This gives a 68% chance that two of the threads had the same seed. All of the Activities generated by those two threads would have colliding IDs.

Comment (pjanotti) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:42]: @tarekgh, @noahfalk is correct: the issue is that a random generator that takes a larger seed is needed to avoid collisions.

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:42]: Ok, will try to look if I can generate a random numbers with 64-bit seeds.

**Comment by tarekgh** [comment]
I have implemented the class to generate a long random numbers. Basically I used the same code/algorithm used in the Random class. I have spent sometime testing it and I am seeing it is good enough (I guess). but if you have any tests you want to try, please let me know and I can try it. Also, I made the objects generating the random numbers be thread static. And here is the perf numbers with the changes:

# Linux

**Before**

|                Method |     Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------------- |---------:|----------:|----------:|-------:|------:|------:|----------:|
| ActivityIdsGeneration | 4.222 us | 0.0294 us | 0.0261 us | 0.0992 |     - |     - |     632 B |

**After**

|                Method |     Mean |   Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------------- |---------:|--------:|--------:|-------:|------:|------:|----------:|
| ActivityIdsGeneration | 299.1 ns | 2.95 ns | 2.61 ns | 0.0625 |     - |     - |     392 B |


# Windows

**Before**

|                Method |     Mean |   Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------------- |---------:|--------:|--------:|-------:|------:|------:|----------:|
| ActivityIdsGeneration | 533.7 ns | 2.91 ns | 2.72 ns | 0.1001 |     - |     - |     632 B |

**After**

|                Method |     Mean |   Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------------- |---------:|--------:|--------:|-------:|------:|------:|----------:|
| ActivityIdsGeneration | 330.7 ns | 0.56 ns | 0.44 ns | 0.0625 |     - |     - |     392 B |

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs]: please have a look and let me know what you think about the last changes.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs
Comment (stephentoub) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:115]: I don't like that we're duplicating all of this code.  https://github.com/dotnet/runtime/issues/26741 is already approved to add NextInt64 to Random... can we just do that and then use it here?  Or just call Next twice?  If for some reason we need to code a custom generator, there are better/faster choices, but it's not clear to me that's needed.

Comment (noahfalk) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:115]: The underlying issue that needs to be solved is that the internal state of Random doesn't start with enough entropy in it because it is constructed from a 32 bit seed. https://github.com/dotnet/runtime/pull/46890#discussion_r556235751. If that problem can be resolved without needing a separate copy of the algorithm thats all the better, but I wasn't counting on that being viable given back-compat expectations.

Comment (noahfalk) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:115]: A little more info:
1. There is no requirement for reproducibility or consistency of implementation over time here, the numbers just need to be collision resistant.
2. DiagnosticSource.dll ships as an OOB package that can be run on downlevel versions of .NET. We could have different PRNG solutions on different versions if we needed to.

Comment (stephentoub) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:115]: > The underlying issue that needs to be solved is that the internal state of Random doesn't start with enough entropy in it because it is constructed from a 32 bit seed

Do we know that just replacing int with long in a copy of the Random code actually meaningfully improves things?  Yes, the seed can now be 64-bit.  Does the rest of the math work out, for a good enough distribution to make your concerns go away?
cc: @tannergooding 

If memory serves, we don't actually like the algorithm used by Random, in that it's both slower and produces lower-quality pseudo-randomness than other algorithms developed in the interim.  We've always just been hesitant to change it due to concerns about compatibility.

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:115]: > Do we know that just replacing int with long in a copy of the Random code actually meaningfully improves things? 

Yes, it does and it has very big difference. From my local testing when using `int`, the period (which the random number can repeat from same generator) is shorter than when using long. When using int this can be less than 100,000. When using long this can happen after 2 billion. 

Also, the algorithm used is called "Additive number generator" described in Knuth book and from the reading which is not tied to specific word size but it is generic and it is expected to get better results if you have bigger word size.

Comment (tannergooding) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:115]: > Does the rest of the math work out, for a good enough distribution to make your concerns go away?

As I just commented on https://github.com/dotnet/runtime/issues/26741, my biggest concern with the `long` random APIs is due to using `double` for scaling the sample to the required range.

I've not looked in depth at this implementation, but it isn't using double anywhere, so my biggest concern is gone 😄

Comment (noahfalk) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:115]: > If memory serves, we don't actually like the algorithm used by Random, in that it's both slower and produces lower-quality pseudo-randomness than other algorithms developed in the interim

I'm glad for us to use any algorithm that we believe is more recommended. I only investigated our existing Random implementation far enough to note that modifying it appeared like a plausible solution. I hadn't made any effort to survey the new state of the art for PRNGs in general.

Comment (stephentoub) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:115]: Just as an example, I translated `Xoshiro256**` from https://en.wikipedia.org/wiki/Xorshift to C#:
```C#
internal class Xoshiro256StarStar
{
	private ulong _s0, _s1, _s2, _s3;

	public unsafe Xoshiro256StarStar()
	{
		while ((_s0 | _s1 | _s2 | _s3) == 0)
		{
			Guid g1 = Guid.NewGuid();
			Guid g2 = Guid.NewGuid();
			ulong* g1p = (ulong*)&g1;
			ulong* g2p = (ulong*)&g2;
			_s0 = *g1p;
			_s1 = *(g1p + 1);
			_s2 = *g2p;
			_s3 = *(g2p + 1);
		}
	}

	private ulong Rol64(ulong x, int k) => (x << k) | (x >> (64 - k));

	public long Next()
	{
		ulong result = Rol64(_s1 * 5, 7) * 9;
		ulong t = _s1 << 17;

		_s2 ^= _s0;
		_s3 ^= _s1;
		_s1 ^= _s2;
		_s0 ^= _s3;

		_s2 ^= t;
		_s3 = Rol64(_s3, 45);

		return (long)result;
	}
}
```
The code is much more concise than the Random code, and it's more than twice as fast:

|         Method |     Mean |     Error |    StdDev |
|--------------- |---------:|----------:|----------:|
| NextXoshiro256 | 3.194 ns | 0.0095 ns | 0.0074 ns |
|   NextRandom64 | 7.644 ns | 0.0916 ns | 0.0857 ns |

while also involving much less allocation for the instance.  It's also reported to have a very good pseudo-random qualities, though I'm admittedly not an expert.

That's all to say, if we're really keen on including a custom generator here, we might as well do it "right".

Comment (stephentoub) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:115]: (That said, I'm still keen on having the .NET Core implementation not use its own generator, and instead rely on System.Random, if at all possible.)

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:115]: > (That said, I'm still keen on having the .NET Core implementation not use its own generator, and instead rely on System.Random, if at all possible.)

What you suggest to do when running on NetFx?

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:115]: > Just as an example, I translated Xoshiro256** from en.wikipedia.org/wiki/Xorshift to C#:

If believe it is more reliable, I can give it a try and test it.

Comment (stephentoub) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:115]: > What you suggest to do when running on NetFx?

Continue using Guid.NewGuid() as we are today.

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:115]: Also, does `Xoshiro256` always return positive values?

Comment (stephentoub) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:115]: > Also, does Xoshiro256 always return positive values?

It covers the full 64-bit space, so if returning long rather than ulong, yes.

But, to be clear, I picked that as an example after about 1 minute of research.  I'm not suggesting it's actually the best option, just highlighting that our System.Random implementation, well, isn't :)

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:115]: > Continue using Guid.NewGuid() as we are today.

What we'll do when running against .NET Core 3.1 on Linux?

Comment (stephentoub) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:115]: > What we'll do when running against .NET Core 3.1/netstandard on Linux?

For .NET Core 3.1, continue using Guid.NewGuid() as we are today.

For .NET Standard, don't we type forward to a built-in implementation when on .NET Core?  Or is it only an OOB?

Comment (stephentoub) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:115]: Again, I'm not dictating here, I'm suggesting we can do the minimum viable for downlevel.  You define what viable is.

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:115]: > For .NET Standard, don't we type forward to a built-in implementation when on .NET Core? Or is it only an OOB?

Yes it is OOB. netstandard.dll don't include any type forward for the types inside this library. 

> For .NET Core 3.1, continue using Guid.NewGuid() as we are today.

I am seeing some customers still care using this library on Core 3.1/5.0. Although this specific issue is not discussed in their scenarios, But the perf is important in general for them.

> Again, I'm not dictating here, I'm suggesting we can do the minimum viable for downlevel. You define what viable is.

Understood. your feedback is very valuable. That is why I was asking more questions to get what I could be missing. 

I want to say I picked the same algorithm used in Random because it has been used for many years and the private implementation I did showing good performance for the scenario we are trying to fix. I already looked at other algorithms which claim it is better (in perf, size, detection...etc.) e.g. https://www.pcg-random.org/ but I hesitated to go this route as I am not sure if the claims are true and if there is any issue with licensing. I don't mind to give such algorithms a try though.

I am inclining to stick with the idea of the custom PRNG regardless which algorithm we use as long as it give us what we need. The reason is, trying to get different behavior on different targeting will make it more complicated for not much benefit. In the future if we start dropping some old targets, we can try to use Random at that time. I am still welcoming any more feedback.

@noahfalk what you think?

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:115]: Just to tell, I gave `Xoshiro256StarStar` a try and seams give a good results as customized Random one.

Comment (noahfalk) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:115]: Given what I've now read online on Xoshiro256** and the positive empirical results I'd be glad to have us use it. Licensing seemed easy as [the code](http://prng.di.unimi.it/xoshiro256starstar.c) is public domain. 

In terms of using a custom PRNG vs. using an in-box one, I agree with your line of thinking Tarek. If we have any scenario where we have to rely on the custom PRNG and we believe the custom one is fast/high quality then it is unclear to me what value is gained by having any alternate implementation option such as using the built-in Random. Potentially it saves some code size but given how compact Xoshiro is that is probably a negligible difference.

There is some (I think fairly small) risk that in the future we'd discover a deficiency in the Xoshiro PRNG algorithm but as long as we avoid making the API public or introducing a new requirement for reproducibility in the IDs we'd have lots of flexibility to change implementation in the future if needed.

**Comment by tarekgh** [comment]
I have changed the used random number generator to the one pointed earlier by @stephentoub. and here is the current perf numbers with that change:

# Linux

**Before**

|                Method |     Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------------- |---------:|----------:|----------:|-------:|------:|------:|----------:|
| ActivityIdsGeneration | 4.067 us | 0.0294 us | 0.0275 us | 0.0992 |     - |     - |     632 B |

**After**

|                Method |     Mean |   Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------------- |---------:|--------:|--------:|-------:|------:|------:|----------:|
| ActivityIdsGeneration | 285.6 ns | 3.50 ns | 3.11 ns | 0.0625 |     - |     - |     392 B |



# Windows 

**Before**

|                Method |     Mean |   Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------------- |---------:|--------:|--------:|-------:|------:|------:|----------:|
| ActivityIdsGeneration | 523.2 ns | 4.18 ns | 3.70 ns | 0.1001 |     - |     - |     632 B |

**After**

|                Method |     Mean |   Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------------- |---------:|--------:|--------:|-------:|------:|------:|----------:|
| ActivityIdsGeneration | 334.6 ns | 1.01 ns | 0.90 ns | 0.0625 |     - |     - |     392 B |

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:61]: Thanks all for the feedback. I have updated the implementation and pated the final perf numbers.

Comment (stephentoub) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:61]: > it is unclear to me what value is gained by having any alternate implementation option such as using the built-in Random

The value is ensuring that we're improving the core platform and not just implementing one-offs here and there because the core implementation isn't currently as good as we'd like it to be.  In general I want us focused on improving the shared code that everyone uses.

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:61]: I think having this one-off here is ok for the sake of benefit when running on Linux with .NET down level versions. In the future we can use the Random when we drop other targets. I am not seeing any conflict with that.

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:61]: @stephentoub would it make sense in the PR https://github.com/dotnet/runtime/pull/47085 to move the `Xoshiro256StarStarImpl` to a common folder and we can include it here? doing that will ensure we still using the shared one and not forking. what you think?

Comment (stephentoub) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:64]: > I think having this one-off here is ok for the sake of benefit when running on Linux with .NET down level versions. In the future we can use the Random when we drop other targets. I am not seeing any conflict with that.

The conflict is when the effort is only put into the one-off and not into fixing the core thing.  If everyone does that, the core thing is never fixed.  And seeing that happen is why I went and did it myself in #47085.

> move the Xoshiro256StarStarImpl and ImplBase to a common folder and we can include it here?

Sure

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:64]: Thanks @stephentoub, if you are going to move `Xoshiro256StarStarImpl` to the common source folder, could you please add the following

```C#
#if ALLOW_PARTIALLY_TRUSTED_CALLERS
        [System.Security.SecuritySafeCriticalAttribute]
#endif
```

on the constructor as it is unsafe and that security attribute is important to have when running on the Ful Framework. 

Please let me know if you prefer merging your PR first and then I move and edit the file myself. Anyway, I have to wait you merge your PR first.

Comment (stephentoub) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:64]: Thanks.  Go ahead with your change.  Once mine is merged in, you can then tweak it as needed to work well for netfx and the other assembly compilation, and dedup.

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:64]: great. could anyone of you approve the current change if there is no any other comment?

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs
Comment (pjanotti) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:34]: @tarekgh one observation: some bits are fixed on the `ulong` captured from the `Guid`. I don't think this is an actual problem since this already has enough bits of entropy but if you want the full 64-bits of entropy you should replace the byte with the fixed bits with a byte without fixed bits from the second-half of the `Guid` (which also has some fixed bits).

**Comment by davidfowl** [comment]
@shirhatti

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:34]: Thanks @pjanotti. That make sense. I have submitted a commit randomizing these bits. please have a look and let me know if you have any comments.

Comment (pjanotti) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:34]: LGTM @tarekgh, thanks for this change.

Comment (pjanotti) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:34]: ops `*(g1p + 1)` also has 2 fixed bits...

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs:34]: I have submitted one more commit to address that. Thanks again.

**Review by noahfalk — APPROVED** [review]
LGTM! I mentioned a few nits but if you are already planning a future PR to converge with Stephen's work feel free to handle it whenever.

**Pull request merged**
